Скачиваем библиотеки

In [51]:
!pip install pandas
!pip install sklearn
!pip install -U scikit-learn
!pip install numpy

  Using cached sklearn-0.0.post12.tar.gz (2.6 kB)
  Installing build dependencies ... done
  Getting requirements to build wheel ... error
  error: subprocess-exited-with-error
  
  × Getting requirements to build wheel did not run successfully.
  │ exit code: 1
  ╰─> [15 lines of output]
      The 'sklearn' PyPI package is deprecated, use 'scikit-learn'
      rather than 'sklearn' for pip commands.
      
      Here is how to fix this error in the main use cases:
      - use 'pip install scikit-learn' rather than 'pip install sklearn'
      - replace 'sklearn' by 'scikit-learn' in your pip requirements files
        (requirements.txt, setup.py, setup.cfg, Pipfile, etc ...)
      - if the 'sklearn' package is used by one of your dependencies,
        it would be great if you take some time to track which package uses
        'sklearn' instead of 'scikit-learn' and report it to their issue tracker
      - as a last resort, set the environment variable
        SKLEARN_ALLOW_DEPRECATED_SK

Импортируем библиотеки

In [53]:
import pandas as pd
from sklearn.preprocessing import StandardScaler
from sklearn.ensemble import RandomForestClassifier
from sklearn.svm import SVC
from sklearn.metrics import accuracy_score, classification_report
from sklearn.pipeline import make_pipeline
from sklearn.metrics import r2_score
from sklearn.neighbors import KNeighborsClassifier
from sklearn.model_selection import train_test_split, GridSearchCV
from sklearn.pipeline import Pipeline
import numpy as np

Загружем данные

In [2]:
full_df = pd.read_csv('full_df.csv')
sample = pd.read_csv('sample.csv')
train = pd.read_csv('train.csv')

Смотрим на них

In [3]:
sample

session_id    target
0               1  0.948255
1               2  0.682483
2               3  0.502855
3               4  0.345556
4               5  0.889428
...           ...       ...
82792       82793  0.585216
82793       82794  0.621074
82794       82795  0.651144
82795       82796  0.319917
82796       82797  0.933516

[82797 rows x 2 columns]

### нормальное время в train

В данном разделе выполняеться Конструирование признаков которая была предоставлена в сессии 2 с данными full_df и там все расписанно

In [4]:
train['time1'] = train['time1'].str.replace('-', ' ')
train['time2'] = train['time2'].str.replace('-', ' ')
train['time3'] = train['time3'].str.replace('-', ' ')
train['time4'] = train['time4'].str.replace('-', ' ')
train['time5'] = train['time5'].str.replace('-', ' ')
train['time6'] = train['time6'].str.replace('-', ' ')
train['time7'] = train['time7'].str.replace('-', ' ')
train['time8'] = train['time8'].str.replace('-', ' ')
train['time9'] = train['time9'].str.replace('-', ' ')
train['time10'] = train['time10'].str.replace('-', ' ')
train

session_id  site1       time1  site2       time2  site3       time3  \
0                2    890  2014 02 22    941  2014 02 22   3847  2014 02 22   
1                3  14769  2013 12 16     39  2013 12 16  14768  2013 12 16   
2                4    782  2014 03 28    782  2014 03 28    782  2014 03 28   
3                5     22  2014 02 28    177  2014 02 28    175  2014 02 28   
4                6    570  2014 03 18     21  2014 03 18    570  2014 03 18   
...            ...    ...         ...    ...         ...    ...         ...   
253555      253557   3474  2013 11 25   3474  2013 11 25    141  2013 11 25   
253556      253558  12727  2013 03 12  12727  2013 03 12   2215  2013 03 12   
253557      253559   2661  2013 09 12  15004  2013 09 12   5562  2013 09 12   
253558      253560    812  2013 12 19    676  2013 12 19    814  2013 12 19   
253559      253561  34942  2014 04 25   1429  2014 04 25  34942  2014 04 25   

        site4       time4  site5  ...       time6  site7       time7  site8  \
0         941  2014 02 22    942  ...  2014 02 22   3847  2014 02 22   3846   
1       14769  2013 12 16     37  ...  2013 12 16  14768  2013 12 16  14768   
2         782  2014 03 28    782  ...  2014 03 28    782  2014 03 28    782   
3         178  2014 02 28    177  ...  2014 02 28    175  2014 02 28    177   
4          21  2014 03 18     21  ...  2014 02 28    175  2014 02 28    177   
...       ...         ...    ...  ...         ...    ...         ...    ...   
253555   2428  2013 11 25    106  ...  2013 11 25   2428  2013 11 25   2428   
253556     38  2013 03 12   2215  ...  2013 03 12  25444  2013 03 12   2215   
253557   5562  2013 09 12   5562  ...  2013 03 12  25444  2013 03 12   2215   
253558     22  2013 12 19     39  ...  2013 12 19    814  2013 12 19    570   
253559     29  2014 04 25     30  ...  2014 04 25   5779  2014 04 25     30   

             time8  site9       time9  site10      time10  target  
0       2014 02 22   1516  2014 02 22    1518  2014 02 22       0  
1       2013 12 16  14768  2013 12 16   14768  2013 12 16       0  
2       2014 03 28    782  2014 03 28     782  2014 03 28       0  
3       2014 02 28    177  2014 02 28     178  2014 02 28       0  
4       2014 02 28    177  2014 02 28     178  2014 02 28       0  
...            ...    ...         ...     ...         ...     ...  
253555  2013 11 25   2428  2013 11 25     148  2013 11 25       0  
253556  2013 03 12     23  2013 03 12      21  2013 03 12       0  
253557  2013 03 12     23  2013 03 12      21  2013 03 12       0  
253558  2013 12 19     22  2013 12 19     570  2013 12 19       0  
253559  2014 04 25     35  2014 04 25      33  2014 04 25       0  

[253560 rows x 22 columns]

In [5]:
train['timef1'] = train['time1'].apply(lambda x: x.split()[0])
train['timea1'] = train['time1'].apply(lambda x: x.split()[1])
train['timef2'] = train['time2'].apply(lambda x: x.split()[0])
train['timea2'] = train['time2'].apply(lambda x: x.split()[1])
train['timef3'] = train['time3'].apply(lambda x: x.split()[0])
train['timea3'] = train['time3'].apply(lambda x: x.split()[1])
train['timef4'] = train['time4'].apply(lambda x: x.split()[0])
train['timea4'] = train['time4'].apply(lambda x: x.split()[1])
train['timef5'] = train['time5'].apply(lambda x: x.split()[0])
train['timea5'] = train['time5'].apply(lambda x: x.split()[1])
train['timef6'] = train['time6'].apply(lambda x: x.split()[0])
train['timea6'] = train['time6'].apply(lambda x: x.split()[1])
train['timef7'] = train['time7'].apply(lambda x: x.split()[0])
train['timea7'] = train['time7'].apply(lambda x: x.split()[1])
train['timef8'] = train['time8'].apply(lambda x: x.split()[0])
train['timea8'] = train['time8'].apply(lambda x: x.split()[1])
train['timef9'] = train['time9'].apply(lambda x: x.split()[0])
train['timea9'] = train['time9'].apply(lambda x: x.split()[1])
train['timef10'] = train['time10'].apply(lambda x: x.split()[0])
train['timea10'] = train['time10'].apply(lambda x: x.split()[1])


train

session_id  site1       time1  site2       time2  site3       time3  \
0                2    890  2014 02 22    941  2014 02 22   3847  2014 02 22   
1                3  14769  2013 12 16     39  2013 12 16  14768  2013 12 16   
2                4    782  2014 03 28    782  2014 03 28    782  2014 03 28   
3                5     22  2014 02 28    177  2014 02 28    175  2014 02 28   
4                6    570  2014 03 18     21  2014 03 18    570  2014 03 18   
...            ...    ...         ...    ...         ...    ...         ...   
253555      253557   3474  2013 11 25   3474  2013 11 25    141  2013 11 25   
253556      253558  12727  2013 03 12  12727  2013 03 12   2215  2013 03 12   
253557      253559   2661  2013 09 12  15004  2013 09 12   5562  2013 09 12   
253558      253560    812  2013 12 19    676  2013 12 19    814  2013 12 19   
253559      253561  34942  2014 04 25   1429  2014 04 25  34942  2014 04 25   

        site4       time4  site5  ... timef6  timea6 timef7  timea7 timef8  \
0         941  2014 02 22    942  ...   2014      02   2014      02   2014   
1       14769  2013 12 16     37  ...   2013      12   2013      12   2013   
2         782  2014 03 28    782  ...   2014      03   2014      03   2014   
3         178  2014 02 28    177  ...   2014      02   2014      02   2014   
4          21  2014 03 18     21  ...   2014      02   2014      02   2014   
...       ...         ...    ...  ...    ...     ...    ...     ...    ...   
253555   2428  2013 11 25    106  ...   2013      11   2013      11   2013   
253556     38  2013 03 12   2215  ...   2013      03   2013      03   2013   
253557   5562  2013 09 12   5562  ...   2013      03   2013      03   2013   
253558     22  2013 12 19     39  ...   2013      12   2013      12   2013   
253559     29  2014 04 25     30  ...   2014      04   2014      04   2014   

        timea8 timef9  timea9 timef10  timea10  
0           02   2014      02    2014       02  
1           12   2013      12    2013       12  
2           03   2014      03    2014       03  
3           02   2014      02    2014       02  
4           02   2014      02    2014       02  
...        ...    ...     ...     ...      ...  
253555      11   2013      11    2013       11  
253556      03   2013      03    2013       03  
253557      03   2013      03    2013       03  
253558      12   2013      12    2013       12  
253559      04   2014      04    2014       04  

[253560 rows x 42 columns]

In [6]:
train = train.drop(['time1', 'time2', 'time3', 'time4', 'time5', 'time6', 'time7', 'time8', 'time9', 'time10'], axis= 1 ,)
train

session_id  site1  site2  site3  site4  site5  site6  site7  site8  \
0                2    890    941   3847    941    942   3846   3847   3846   
1                3  14769     39  14768  14769     37     39  14768  14768   
2                4    782    782    782    782    782    782    782    782   
3                5     22    177    175    178    177    178    175    177   
4                6    570     21    570     21     21    178    175    177   
...            ...    ...    ...    ...    ...    ...    ...    ...    ...   
253555      253557   3474   3474    141   2428    106   2428   2428   2428   
253556      253558  12727  12727   2215     38   2215     23  25444   2215   
253557      253559   2661  15004   5562   5562   5562     23  25444   2215   
253558      253560    812    676    814     22     39    812    814    570   
253559      253561  34942   1429  34942     29     30  34942   5779     30   

        site9  ...  timef6  timea6 timef7 timea7 timef8 timea8 timef9 timea9  \
0        1516  ...    2014      02   2014     02   2014     02   2014     02   
1       14768  ...    2013      12   2013     12   2013     12   2013     12   
2         782  ...    2014      03   2014     03   2014     03   2014     03   
3         177  ...    2014      02   2014     02   2014     02   2014     02   
4         177  ...    2014      02   2014     02   2014     02   2014     02   
...       ...  ...     ...     ...    ...    ...    ...    ...    ...    ...   
253555   2428  ...    2013      11   2013     11   2013     11   2013     11   
253556     23  ...    2013      03   2013     03   2013     03   2013     03   
253557     23  ...    2013      03   2013     03   2013     03   2013     03   
253558     22  ...    2013      12   2013     12   2013     12   2013     12   
253559     35  ...    2014      04   2014     04   2014     04   2014     04   

       timef10 timea10  
0         2014      02  
1         2013      12  
2         2014      03  
3         2014      02  
4         2014      02  
...        ...     ...  
253555    2013      11  
253556    2013      03  
253557    2013      03  
253558    2013      12  
253559    2014      04  

[253560 rows x 32 columns]

In [7]:
train['time_s1'] = train['timef1'] + train['timea1']
train['time_s2'] = train['timef2'] + train['timea2']
train['time_s3'] = train['timef3'] + train['timea3']
train['time_s4'] = train['timef4'] + train['timea4']
train['time_s5'] = train['timef5'] + train['timea5']
train['time_s6'] = train['timef6'] + train['timea6']
train['time_s7'] = train['timef7'] + train['timea7']
train['time_s8'] = train['timef8'] + train['timea8']
train['time_s9'] = train['timef9'] + train['timea9']
train['time_s10'] = train['timef10'] + train['timea10']
train

session_id  site1  site2  site3  site4  site5  site6  site7  site8  \
0                2    890    941   3847    941    942   3846   3847   3846   
1                3  14769     39  14768  14769     37     39  14768  14768   
2                4    782    782    782    782    782    782    782    782   
3                5     22    177    175    178    177    178    175    177   
4                6    570     21    570     21     21    178    175    177   
...            ...    ...    ...    ...    ...    ...    ...    ...    ...   
253555      253557   3474   3474    141   2428    106   2428   2428   2428   
253556      253558  12727  12727   2215     38   2215     23  25444   2215   
253557      253559   2661  15004   5562   5562   5562     23  25444   2215   
253558      253560    812    676    814     22     39    812    814    570   
253559      253561  34942   1429  34942     29     30  34942   5779     30   

        site9  ...  time_s1  time_s2 time_s3 time_s4 time_s5 time_s6 time_s7  \
0        1516  ...   201402   201402  201402  201402  201402  201402  201402   
1       14768  ...   201312   201312  201312  201312  201312  201312  201312   
2         782  ...   201403   201403  201403  201403  201403  201403  201403   
3         177  ...   201402   201402  201402  201402  201402  201402  201402   
4         177  ...   201403   201403  201403  201403  201403  201402  201402   
...       ...  ...      ...      ...     ...     ...     ...     ...     ...   
253555   2428  ...   201311   201311  201311  201311  201311  201311  201311   
253556     23  ...   201303   201303  201303  201303  201303  201303  201303   
253557     23  ...   201309   201309  201309  201309  201309  201303  201303   
253558     22  ...   201312   201312  201312  201312  201312  201312  201312   
253559     35  ...   201404   201404  201404  201404  201404  201404  201404   

       time_s8 time_s9 time_s10  
0       201402  201402   201402  
1       201312  201312   201312  
2       201403  201403   201403  
3       201402  201402   201402  
4       201402  201402   201402  
...        ...     ...      ...  
253555  201311  201311   201311  
253556  201303  201303   201303  
253557  201303  201303   201303  
253558  201312  201312   201312  
253559  201404  201404   201404  

[253560 rows x 42 columns]

In [8]:
train = train.drop(['timef1', 'timef2', 'timef3', 'timef4', 'timef5', 'timef6', 'timef7', 'timef8', 'timef9', 'timef10','timea1', 'timea2', 'timea3', 'timea4', 'timea5', 'timea6', 'timea7', 'timea8', 'timea9', 'timea10'], axis= 1 ,)
train

session_id  site1  site2  site3  site4  site5  site6  site7  site8  \
0                2    890    941   3847    941    942   3846   3847   3846   
1                3  14769     39  14768  14769     37     39  14768  14768   
2                4    782    782    782    782    782    782    782    782   
3                5     22    177    175    178    177    178    175    177   
4                6    570     21    570     21     21    178    175    177   
...            ...    ...    ...    ...    ...    ...    ...    ...    ...   
253555      253557   3474   3474    141   2428    106   2428   2428   2428   
253556      253558  12727  12727   2215     38   2215     23  25444   2215   
253557      253559   2661  15004   5562   5562   5562     23  25444   2215   
253558      253560    812    676    814     22     39    812    814    570   
253559      253561  34942   1429  34942     29     30  34942   5779     30   

        site9  ...  time_s1  time_s2 time_s3 time_s4 time_s5 time_s6 time_s7  \
0        1516  ...   201402   201402  201402  201402  201402  201402  201402   
1       14768  ...   201312   201312  201312  201312  201312  201312  201312   
2         782  ...   201403   201403  201403  201403  201403  201403  201403   
3         177  ...   201402   201402  201402  201402  201402  201402  201402   
4         177  ...   201403   201403  201403  201403  201403  201402  201402   
...       ...  ...      ...      ...     ...     ...     ...     ...     ...   
253555   2428  ...   201311   201311  201311  201311  201311  201311  201311   
253556     23  ...   201303   201303  201303  201303  201303  201303  201303   
253557     23  ...   201309   201309  201309  201309  201309  201303  201303   
253558     22  ...   201312   201312  201312  201312  201312  201312  201312   
253559     35  ...   201404   201404  201404  201404  201404  201404  201404   

       time_s8 time_s9 time_s10  
0       201402  201402   201402  
1       201312  201312   201312  
2       201403  201403   201403  
3       201402  201402   201402  
4       201402  201402   201402  
...        ...     ...      ...  
253555  201311  201311   201311  
253556  201303  201303   201303  
253557  201303  201303   201303  
253558  201312  201312   201312  
253559  201404  201404   201404  

[253560 rows x 22 columns]

# 3.1 Сравнение нескольких алгоритмов на сессиях

Данные разделенны 20 на 80 ведь это дает наилучший результат для обучения модели

В данном коде используеться метрика r2_score ведь она на мой взгляд дает лучший результат чем другие метрики и даст наилучший результат в данной задаче

<math xmlns="http://www.w3.org/1998/Math/MathML" display="block">
  <msup>
    <mi>R</mi>
    <mn>2</mn>
  </msup>
  <mo stretchy="false">(</mo>
  <mi>y</mi>
  <mo>,</mo>
  <mrow data-mjx-texclass="ORD">
    <mover>
      <mi>y</mi>
      <mo stretchy="false">^</mo>
    </mover>
  </mrow>
  <mo stretchy="false">)</mo>
  <mo>=</mo>
  <mn>1</mn>
  <mo>&#x2212;</mo>
  <mfrac>
    <mrow>
      <munderover>
        <mo data-mjx-texclass="OP">&#x2211;</mo>
        <mrow data-mjx-texclass="ORD">
          <mi>i</mi>
          <mo>=</mo>
          <mn>1</mn>
        </mrow>
        <mrow data-mjx-texclass="ORD">
          <msub>
            <mi>n</mi>
            <mrow data-mjx-texclass="ORD">
              <mi>s</mi>
              <mi>a</mi>
              <mi>m</mi>
              <mi>p</mi>
              <mi>l</mi>
              <mi>e</mi>
              <mi>s</mi>
            </mrow>
          </msub>
        </mrow>
      </munderover>
      <mo stretchy="false">(</mo>
      <msub>
        <mi>y</mi>
        <mi>i</mi>
      </msub>
      <mo>&#x2212;</mo>
      <msub>
        <mrow data-mjx-texclass="ORD">
          <mover>
            <mi>y</mi>
            <mo stretchy="false">^</mo>
          </mover>
        </mrow>
        <mi>i</mi>
      </msub>
      <msup>
        <mo stretchy="false">)</mo>
        <mn>2</mn>
      </msup>
    </mrow>
    <mrow>
      <munderover>
        <mo data-mjx-texclass="OP">&#x2211;</mo>
        <mrow data-mjx-texclass="ORD">
          <mi>i</mi>
          <mo>=</mo>
          <mn>1</mn>
        </mrow>
        <mrow data-mjx-texclass="ORD">
          <msub>
            <mi>n</mi>
            <mrow data-mjx-texclass="ORD">
              <mi>s</mi>
              <mi>a</mi>
              <mi>m</mi>
              <mi>p</mi>
              <mi>l</mi>
              <mi>e</mi>
              <mi>s</mi>
            </mrow>
          </msub>
        </mrow>
      </munderover>
      <mrow data-mjx-texclass="ORD">
        <mo stretchy="false">(</mo>
        <msub>
          <mi>y</mi>
          <mi>i</mi>
        </msub>
        <mo>&#x2212;</mo>
        <msub>
          <mrow data-mjx-texclass="ORD">
            <mover>
              <mi>y</mi>
              <mo stretchy="false">&#xAF;</mo>
            </mover>
          </mrow>
          <mi>i</mi>
        </msub>
        <msup>
          <mo stretchy="false">)</mo>
          <mn>2</mn>
        </msup>
      </mrow>
    </mrow>
  </mfrac>
</math>

## 1.Обоснование выбора модели: SVM

Support Vector Machine (SVM): SVM хорошо работает для задач классификации, особенно в случае, когда данные не очень объемные, и есть ярко выраженные разделители между классами. SVM также позволяет легко справляться с нелинейными зависимостями с помощью ядерных функций.

Pipeline с StandardScaler: StandardScaler используется для стандартизации данных, что особенно важно для методов, таких как SVM, которые чувствительны к масштабу признаков.

Простота и эффективность: SVM - относительно простой и эффективный метод классификации, особенно когда у вас есть небольшой объем данных.


Задаем в переменные x и у свои данные

In [9]:
X = train.drop('target', axis=1)
y = train['target']

Разделим данные на тренировочный и тестовый наборы

In [10]:
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

Создадим pipeline с StandardScaler и SVM

In [22]:
model = make_pipeline(StandardScaler(), SVC())

Обучим модель на тренировочных данных

In [23]:
model.fit(X_train, y_train)

Pipeline(steps=[('standardscaler', StandardScaler()), ('svc', SVC())])

Предскажем метки классов для тестового набора

In [24]:
y_pred = model.predict(X_test)

Оценка производительности/эффективности модели

In [27]:
# метрика r2
r2_score(model.predict(X_test), y_test)

0.0

Оценим точность модели

In [28]:
accuracy = accuracy_score(y_test, y_pred)
print(f'Точность модели: {accuracy}')
classification_rep = classification_report(y_test, y_pred)
print('Отчет по классификации:\n', classification_rep)

Точность модели: 0.9909094494399747
Отчет по классификации:
               precision    recall  f1-score   support

           0       0.99      1.00      1.00     50251
           1       0.00      0.00      0.00       461

    accuracy                           0.99     50712
   macro avg       0.50      0.50      0.50     50712
weighted avg       0.98      0.99      0.99     50712



/home/admin1/Desktops/Desktop1/Сессия 3/.venv/lib/python3.11/site-packages/sklearn/metrics/_classification.py:1509: UndefinedMetricWarning: Precision is ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, f"{metric.capitalize()} is", len(result))
/home/admin1/Desktops/Desktop1/Сессия 3/.venv/lib/python3.11/site-packages/sklearn/metrics/_classification.py:1509: UndefinedMetricWarning: Precision is ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, f"{metric.capitalize()} is", len(result))
/home/admin1/Desktops/Desktop1/Сессия 3/.venv/lib/python3.11/site-packages/sklearn/metrics/_classification.py:1509: UndefinedMetricWarning: Precision is ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(avera

## 2.Обоснование выбора модели: RandomForest

Random Forest (Случайный лес): Эта модель хорошо обрабатывает разнообразные типы данных, автоматически обрабатывает пропущенные значения, устойчива к переобучению и не требует масштабирования признаков. Также, она способна работать с большим количеством признаков, что может быть важным при анализе данных о дорожных ситуациях.


Задаем в переменные x и у свои данные

In [42]:
X = train.drop('target', axis=1)
y = train['target']

Разделим данные на тренировочный и тестовый наборы

In [43]:
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

Обучим модель на тренировочных данных

In [44]:
model = RandomForestClassifier(n_estimators=100, random_state=42)
model.fit(X_train, y_train)

RandomForestClassifier(random_state=42)

Предскажем метки классов для тестового набора

In [45]:
y_pred = model.predict(X_test)

Оценка производительности/эффективности модели

In [46]:
# метрика r2
r2_score(model.predict(X_test), y_test)

-2.9497698613412884

Оценим точность модели

In [47]:
accuracy = accuracy_score(y_test, y_pred)
print(f'Точность модели: {accuracy}')

Точность модели: 0.991994005363622


## 3.Обоснование выбора модели: Knn

Метод k-ближайших соседей (k-NN) выбирается в различных сценариях из-за своей простоты и интуитивной идеи

Задаем в переменные x и у свои данные


Задаем в переменные x и у свои данные

In [35]:
X = train.drop('target', axis=1)
y = train['target']

Разделим данные на тренировочный и тестовый наборы

In [36]:
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

Обучим модель на тренировочных данных

In [38]:
knn_classifier = KNeighborsClassifier(n_neighbors=3)
knn_classifier.fit(X_train, y_train)

KNeighborsClassifier(n_neighbors=3)

Предскажем метки классов для тестового набора

In [39]:
y_pred = knn_classifier.predict(X_test)

In [40]:
# метрика r2
r2_score(model.predict(X_test), y_test)

-2.9497698613412884

Оценим точность модели

In [41]:
accuracy = accuracy_score(y_test, y_pred)
print(f"Accuracy: {accuracy}")

Accuracy: 0.9904361886732923


## Лучшей моделью оказалась RandomForest с результатом 0.991994005363622

# 3.2 Гиперпараметрическая оптимизация модели

В данном коде производим мы ищем оптимальные гиперпараметры для модели

In [48]:

# Создание конвейера с шагами нормализации и классификации
pipeline = Pipeline([
    ('scaler', StandardScaler()),  # Нормализация данных
    ('rf', model)  # Модель RandomForest
])

# Определение параметров для Grid Search
param_grid = {
    'rf__n_estimators': [10, 50, 100],
    'rf__max_depth': [ 10, 20, 30],
    'rf__min_samples_split': [2, 5, 10],
    'rf__min_samples_leaf': [1, 2, 4]
}

# Создание объекта Grid Search
grid_search = GridSearchCV(pipeline, param_grid=param_grid, cv=5, scoring='accuracy')

# Обучение модели с использованием Grid Search
grid_search.fit(X_train, y_train)

# Вывод оптимальных параметров
print("Оптимальные параметры:", grid_search.best_params_)

# Оценка производительности модели на тестовых данных
accuracy = grid_search.score(X_test, y_test)
print(f"Точность на тестовых данных: {accuracy:.2f}")

KeyboardInterrupt: 

# 3.3 Прогноз классификации для тестовой выборки

Сохраняем прогнозы

In [54]:
def write_to_submission_file(predicted_labels, out_file,
                             target='target', index_label="session_id"):
    predicted_df = pd.DataFrame(predicted_labels,
                                index = np.arange(1, predicted_labels.shape[0] + 1),
                                columns=[target])
    predicted_df.to_csv(out_file, index_label=index_label)

In [55]:
write_to_submission_file(y_pred, 'sample_sub.csv')

# 3.4 Подготовка отчета

В результате этой сессии была выполнена почти вся работа

3.1 Сравнение нескольких алгоритмов на сессиях
- Было выполнено разбиение выборки обучающую и отложенную с обьяснениями.Было проведено обучение
моделей, основанных на различных алгоритмах. Был сделан прогноз в виде предсказанных
вероятностей того, что сессия является сессией злоумышленника на отложенной выборке.
Была выполнена оценка моделей разной степени сложности в соответствии с выбранной метрикой,
предусмотренно распараллеливание процессов. Была выбрана модель, показавшая лучшее качество.(RandomForest с результатом ~0.992).
Обьяснения выбора модели 
Support Vector Machine (SVM): SVM хорошо работает для задач классификации, особенно в случае, когда данные не очень объемные, и есть ярко выраженные разделители между классами. SVM также позволяет легко справляться с нелинейными зависимостями с помощью ядерных функций.
Pipeline с StandardScaler: StandardScaler используется для стандартизации данных, что особенно важно для методов, таких как SVM, которые чувствительны к масштабу признаков.
Простота и эффективность: SVM - относительно простой и эффективный метод классификации, особенно когда у вас есть небольшой объем данных.
Random Forest (Случайный лес): Эта модель хорошо обрабатывает разнообразные типы данных, автоматически обрабатывает пропущенные значения, устойчива к переобучению и не требует масштабирования признаков. Также, она способна работать с большим количеством признаков, что может быть важным при анализе данных о дорожных ситуациях.
Метод k-ближайших соседей (k-NN) выбирается в различных сценариях из-за своей простоты и интуитивной идеи

3.2 Гиперпараметрическая оптимизация модели
- После выбора модели была оптимизациия её под решаемую задачу, настраивая
гиперпараметры.Была выполнена оценка данной модели после
настройки гиперпараметров.
Улучшение модели происходит с помошью GridSearch и нормализации данных

3.3 Прогноз классификации для тестовой выборки
- Был сделан прогноз в виде предсказанных вероятностей отнесенных к классу 1 для тестовой
выборки с помощью полученной модели, обученной уже на всей обучающей выборке. Ответы были
записанны в файл в соответствии с шаблоном sample_sub.csv

3.4 Подготовка отчета
- Был сделан отчет о проделанной работе по итогам сессии, в котором
представлены результаты, выводы и обоснования выбора по каждому разделу задания.
Результаты работы состоят из отчетов в формате .html и исходников с возможностью
перекомпиляции. Архива Data.zip который содержит все результаты выполнения модуля, а также
все необходимые файлы для запуска и проверки участков кода. В файле Readme.txt расписанно
содержимое результирующих файлов архива Data.zip.

In [ ]:
train.to_csv('train_new.csv',index=False)